src/mpi/coll/iallgather/iallgather_tsp_ring_algos.h

```diff
@@ -0,0 +1,195 @@
+/* -*- Mode: C; c-basic-offset:4 ; indent-tabs-mode:nil ; -*- */
+/*
+ *  (C) 2006 by Argonne National Laboratory.
+ *      See COPYRIGHT in top-level directory.
+ *
+ *  Portions of this code were written by Intel Corporation.
+ *  Copyright (C) 2011-2017 Intel Corporation.  Intel provides this material
+ *  to Argonne National Laboratory subject to Software Grant and Corporate
+ *  Contributor License Agreement dated February 8, 2012.
+ */
+
+/* Header protection (i.e., IALLGATHER_TSP_RING_ALGOS_H_INCLUDED) is
+ * intentionally omitted since this header might get included multiple
+ * times within the same .c file. */
+
+#include "tsp_namespace_def.h"
+
+/* Routine to schedule a ring based allgather */
+#undef FUNCNAME
+#define FUNCNAME MPIR_TSP_Iallgather_sched_intra_ring
+#undef FCNAME
+#define FCNAME MPL_QUOTE(FUNCNAME)
+int MPIR_TSP_Iallgather_sched_intra_ring(const void *sendbuf, int sendcount,
+                                         MPI_Datatype sendtype, void *recvbuf,
+                                         int recvcount, MPI_Datatype recvtype,
+                                         MPIR_Comm * comm, MPIR_TSP_sched_t * sched)
+{
+    int mpi_errno = MPI_SUCCESS;
+    int i, src, dst, copy_dst;
+    /* Temporary buffers to execute the ring algorithm */
+    void *buf1, *buf2, *data_buf, *rbuf, *sbuf;
+    int vtcs[3], send_id[3], recv_id[3], dtcopy_id[3];
+    int nvtcs;
+
+    int size = MPIR_Comm_size(comm);
+    int rank = MPIR_Comm_rank(comm);
+    int is_inplace = (sendbuf == MPI_IN_PLACE);
+    int tag;
+
+    size_t recvtype_lb, recvtype_extent;
+    size_t sendtype_lb, sendtype_extent;
+    size_t sendtype_true_extent, recvtype_true_extent;
+
+    MPIR_FUNC_VERBOSE_STATE_DECL(MPID_STATE_MPIR_TSP_IALLGATHER_SCHED_INTRA_RING);
+    MPIR_FUNC_VERBOSE_ENTER(MPID_STATE_MPIR_TSP_IALLGATHER_SCHED_INTRA_RING);
+
+    /* Find out the buffer which has the send data and point data_buf to it */
+    if (is_inplace) {
+        sendcount = recvcount;
+        sendtype = recvtype;
+        data_buf = (char *) recvbuf;
+    } else
+        data_buf = (char *) sendbuf;
```

Comment:
`data_buf` is only used during `localcopy` and we have to switch on `is_inplace` there anyway. Please move it to below.